algorithms.py

```diff
@@ -604,21 +802,67 @@ def hudson_recombination_event(self):
                 y.next.prev = z
             y.next = None
             y.right = k
-            self.L.increment(y.index, k - z.right)
-            lhs_tail = y
+            self.L[label].increment(y.index, k - z.right)
         else:
             # split the link between x and y.
             x.next = None
             y.prev = None
             z = y
             lhs_tail = x
-        self.L.set_value(z.index, z.right - z.left - 1)
-        self.P[z.population].add(z)
+        z.label = label
+        self.L[label].set_value(z.index, z.right - z.left - 1)
+        self.P[z.population].add(z, label)
         if self.full_arg:
             self.store_node(lhs_tail.population, flags=msprime.NODE_IS_RE_EVENT)
             self.store_arg_edges(lhs_tail)
             self.store_node(z.population, flags=msprime.NODE_IS_RE_EVENT)
             self.store_arg_edges(z)
+        ret = None
+        if return_heads:
+            if x is None:
```

Comment:
We can simplify this a bit now:
```suggestion
                    x = lhs_tail
                    # Seek back to the head of the x chain
￼	            while x.prev is not None:
￼	                x = x.prev
￼	            ret = x, z
```

Reply:
Not sure how these suggested changes things work, so that won't be valid code. Basically, we don't need the ``if x is None`` check any more.

Reply:
so i'm probably missing something here, but i think we still need the `if x is None:` check. From the bit of testing I've done the code is working...

Reply:
No worries, this wasn't very clear. I'll sort it out and merge.

I'll be force pushing to your master branch though, so beware it'll be mucked up. (It would be simpler if you used a topic branch rather than master, as you can just delete these after they're merged)

Reply:
force push away my friend. will use topic branches from now on-- getting used to them with the stdpopsim repo!